rules/opentitan.bzl

```diff
@@ -0,0 +1,249 @@
+            **kwargs
+        )
+        targets.append(":" + devname + "_elf")
+        elf_to_binary(
+            name = devname + "_elf",
+            srcs = [devname],
+            format = "elf32-little",
+            suffix = "elf",
+            platform = platform,
+        )
+
+        if output_bin:
+            targets.append(":" + devname + "_bin")
+            elf_to_binary(
+                name = devname + "_bin",
+                srcs = [devname],
+                platform = platform,
+            )
+
+        if output_disassembly:
+            targets.append(":" + devname + "_dis")
+            elf_to_disassembly(
+                name = devname + "_dis",
+                srcs = [devname],
+                platform = platform,
+            )
+        if output_scrambled:
+            targets.append(":" + devname + "_scr")
+            elf_to_scrambled(
+                name = devname + "_scr",
+                srcs = [devname],
+                platform = platform,
+            )
+
+    native.filegroup(
+        name = name,
+        srcs = targets,
+    )
+
+def opentitan_functest(
+        name,
+        platform = OPENTITAN_PLATFORM,
+        per_device_deps = {
+            "verilator": ["//sw/device/lib/arch:sim_verilator"],
+        },
+        **kwargs):
+    """A helper macro for generating OpenTitan functional tests.
+    This macro is mostly a wrapper around opentitan_binary, but creates
+    testing artifacts for each of the keys in `per_device_deps`.
+    The testing artifacts are then given to an `sh_test` rule which
+    coordinates performing the test under verilator by way of opentitantool.
+    Args:
+      @param name: The name of this rule.
+      @param platform: The target platform for the artifacts.
+      @param per_device_deps: The deps for each of the execution environments.
+      @param **kwargs: Arguments to forward to `opentitan_binary`.
+    """
+
+    opentitan_binary(
+        name = name + "_prog",
+        platform = platform,
+        per_device_deps = per_device_deps,
+        output_bin = False,
+        output_disassembly = False,
+        **kwargs
+    )
+
+    native.sh_test(
+        name = name,
+        srcs = ["//util:opentitantool_test_runner.sh"],
+        args = [
+            "--tool=$(location //sw/host/opentitantool)",
+            "--verilator-bin=$(location //prebuilt:Vchip_earlgrey_verilator)",
+            "--verilator-rom=$(location //sw/device/boot_rom:boot_rom_verilator_scr)",
+            "--verilator-flash=$(location {}_prog_verilator_elf)".format(name),
+            "--verilator-otp=$(location //:otp_image_verilator)",
+        ],
+        data = [
+            "//sw/device/boot_rom:boot_rom_verilator_scr",
+            "{}_prog_verilator_elf".format(name),
+            "//sw/host/opentitantool",
+            "//prebuilt:Vchip_earlgrey_verilator",
+            "//:otp_image_verilator",
+        ],
+    )
```

Comment:
I think we should omit this for now.  We need to get opentitantool building with bazel and work out a mechanism for dispatching tests among the various supported backends.